app/src/main/java/fr/neamar/kiss/utils/FuzzyScore.java

```diff
@@ -0,0 +1,291 @@
+	/**
+	 * bonus if match is uppercase and prev is lower
+	 */
+	private int camel_bonus;
+	/**
+	 * penalty applied for every letter in str before the first match
+	 */
+	private int leading_letter_penalty;
+	/**
+	 * maximum penalty for leading letters
+	 */
+	private int max_leading_letter_penalty;
+	/**
+	 * penalty for every letter that doesn't matter
+	 */
+	private int unmatched_letter_penalty;
+
+	public FuzzyScore()
+	{
+		super();
+		adjacency_bonus = 5;
+		separator_bonus = 10;
+		camel_bonus = 10;
+		leading_letter_penalty = -3;
+		max_leading_letter_penalty = -9;
+		unmatched_letter_penalty = -1;
+	}
+
+	public void setAdjacencyBonus( int adjacency_bonus )
+	{
+		this.adjacency_bonus = adjacency_bonus;
+	}
+
+	public void setSeparatorBonus( int separator_bonus )
+	{
+		this.separator_bonus = separator_bonus;
+	}
+
+	public void setCamelBonus( int camel_bonus )
+	{
+		this.camel_bonus = camel_bonus;
+	}
+
+	public void setLeadingLetterPenalty( int leading_letter_penalty )
+	{
+		this.leading_letter_penalty = leading_letter_penalty;
+	}
+
+	public void setMaxLeadingLetterPenalty( int max_leading_letter_penalty )
+	{
+		this.max_leading_letter_penalty = max_leading_letter_penalty;
+	}
+
+	public void setUnmatchedLetterPenalty( int unmatched_letter_penalty )
+	{
+		this.unmatched_letter_penalty = unmatched_letter_penalty;
+	}
+
+	public static class MatchInfo
+	{
+		/**
+		 * higher is better match. Value has no intrinsic meaning. Range varies with pattern.
+		 * Can only compare scores with same search pattern.
+		 */
+		public int    score;
+		//		/**
+//		 * input str with matched characters marked between {}
+//		 */
+//		public String formattedString;
+		public ArrayList<Integer> matchedIndices;
+
+		public List<Pair<Integer, Integer>> getMatchedSequences()
+		{
+			// compute pair match indices
+			List<Pair<Integer, Integer>> positions = new ArrayList<>( matchedIndices.size() );
+			int                          start     = this.matchedIndices.get( 0 );
+			int                          end       = start + 1;
+			for( int i = 1; i < this.matchedIndices.size(); i += 1 )
+			{
+				if( end == this.matchedIndices.get( i ) )
+				{
+					end += 1;
+				}
+				else
+				{
+					positions.add( new Pair<>( start, end ) );
+					start = this.matchedIndices.get( i );
+					end = start + 1;
+				}
+			}
+			positions.add( new Pair<>( start, end ) );
+			return positions;
+		}
+	}
+
+	/**
+	 * Returns true if each character in pattern is found sequentially within str
+	 *
+	 * @param pattern
+	 * @param str
+	 * @return true if each character in pattern is found sequentially within str
+	 */
+	public boolean match( String pattern, String str )
+	{
+		int patternIdx    = 0;
+		int strIdx        = 0;
+		int patternLength = pattern.length();
+		int strLength     = str.length();
+
+		while( patternIdx != patternLength && strIdx != strLength )
+		{
+			int patternChar = Character.toLowerCase( pattern.codePointAt( patternIdx ) );

```

Comment:
can we send the Pattern when we create a new Fuzzy object, and then store an int[] of all lower case codePoint for the pattern instead of regenerating every time?